l2geth/core/vm/evm.go

```diff
@@ -325,40 +371,47 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 
 	if UsingOVM {
 		// OVM_ENABLED
-
 		if evm.depth == 0 {
 			// We're back at the root-level message call, so we'll need to modify the return data
 			// sent to us by the OVM_ExecutionManager to instead be the intended return data.
 
-			if len(ret) >= 96 {
-				// We expect that EOA contracts return at least 96 bytes of data, where the first
-				// 32 bytes are the boolean success value and the next 64 bytes are unnecessary
-				// ABI encoding data. The actual return data starts at the 96th byte and can be
-				// empty.
-				success := ret[:32]
-				ret = ret[96:]
-
-				if !bytes.Equal(success, AbiBytesTrue) && !bytes.Equal(success, AbiBytesFalse) {
-					// If the first 32 bytes not either are the ABI encoding of "true" or "false",
-					// then the user hasn't correctly ABI encoded the result. We return the null
-					// hex string as a default here (an annoying default that would convince most
-					// people to just use the standard form).
-					ret = common.FromHex("0x")
-				} else if bytes.Equal(success, AbiBytesFalse) {
-					// If the first 32 bytes are the ABI encoding of "false", then we need to add an
-					// artificial error that represents the revert.
-					err = errExecutionReverted
+			// Attempt to decode the returndata as as ExecutionManager.run when
+			// it is not an `eth_call` and as ExecutionManager.simulateMessage
+			// when it is an `eth_call`. If the data is not decodable as ABI
+			// encoded bytes, then return nothing. If the data is able to be
+			// decoded as bytes, then attempt to decode as (bool, bytes)
+			isDecodable := true
+			run := runReturnData{}
+			switch evm.Context.EthCallSender {
+			case nil:
+				if err := evm.Context.OvmExecutionManager.ABI.Unpack(&run, "run", ret); err != nil {
+					isDecodable = false
+				}
+			default:
+				if err := evm.Context.OvmExecutionManager.ABI.Unpack(&run, "simulateMessage", ret); err != nil {
+					isDecodable = false
```

Comment:
Maybe there is a reason to leave it, but notably these code paths are duplicate in practice with the return types currently laid out, and either would work in both cases.

Reply:
I am not super opinionated on leaving both or using a custom coder that has the exact methods, my intention was to be explicit

Reply:
This would be cleaner to use the `codec` and unpack a method that will abi decode bytes

Reply:
IMO better to use both since they're separate functions and may change.

Reply:
We need to standardize these 2 methods such that they are the same, they should never change